src/array_api_typing/_array.py

```diff
@@ -67,10 +74,131 @@ def dtype(self, /) -> DTypeT_co:
         ...
 
 
+class HasDevice(Protocol):
+    """Protocol for array classes that have a device attribute."""
+
+    @property
+    def device(self) -> object:  # TODO: more specific type
+        """Hardware device the array data resides on."""
+        ...
+
+
+class HasMatrixTranspose(Protocol):
+    """Protocol for array classes that have a matrix transpose attribute."""
+
+    @property
+    def mT(self) -> Self:  # noqa: N802
+        """Transpose of a matrix (or a stack of matrices).
+
+        If an array instance has fewer than two dimensions, an error should be
+        raised.
+
+        Returns:
+            Self: array whose last two dimensions (axes) are permuted in reverse
+                order relative to original array (i.e., for an array instance
+                having shape `(..., M, N)`, the returned array must have shape
+                `(..., N, M))`.  The returned array must have the same data type
+                as the original array.
+
+        """
+        ...
+
+
+class HasNDim(Protocol):
+    """Protocol for array classes that have a number of dimensions attribute."""
+
+    @property
+    def ndim(self) -> int:
+        """Number of array dimensions (axes).
+
+        Returns:
+            int: number of array dimensions (axes).
+
+        """
+        ...
+
+
+class HasShape(Protocol):
+    """Protocol for array classes that have a shape attribute."""
+
+    @property
+    def shape(self) -> tuple[int | None, ...]:
+        """Shape of the array.
+
+        Returns:
+            tuple[int | None, ...]: array dimensions. An array dimension must be None
+                if and only if a dimension is unknown.
+
+        Notes:
+            For array libraries having graph-based computational models, array
+            dimensions may be unknown due to data-dependent operations (e.g.,
+            boolean indexing; `A[:, B > 0]`) and thus cannot be statically
+            resolved without knowing array contents.
+
+        """
+        ...
+
+
+class HasSize(Protocol):
```

Comment:
What's the use-case of this one? Is there anything this can help with, that `HasShape` can't?
Put differently; should we make this public API or not?

Reply:
```python
def is_sized(obj: Any, /) -> TypeGuard[HasSize]: ...
```

Reply:
IDK personally, but the Array API is pretty sparing.
I could see using something like:

```python

def check_min_size(obj: object) -> bool:
    if not is_sized(obj) elif obj.size is None:
        raise Exception
    return obj.size > 10
```

Reply:
Hm yea I guess, but that's probably pretty niche, right? 

I guess I just feel like we should first start with fixing the big typing problems that the array-api libs are facing, and I don't consider this to be one of those. That's why I opened #22, so that we can get a better understanding of what those problems actually are. And right now it just feels like we're trying to mimick the runtime API, and blindly hope that it will somehow be helpful, without knowing how exactly.

Reply:
Not disagreeing that there are other big issues to solve, but people also want and need the more straightforward static Array API. I've had *many* people ask me about doing straightforward things like

```python
def square[T=Array](x: T) -> T: ...
```

So yeah, I do think there's a lot of value in starting to fill out the static equivalent to the runtime API.
We know it will be useful. And how.

Reply:
> But I'm trying to prevent is situations where there are multiple non-obvious ways of achieving the same result. And `HasSize` seems to me like it could have a lot of overlap with `HasShape`.

This library isn't the venue for changing the Array API...

Reply:
> > But I'm trying to prevent is situations where there are multiple non-obvious ways of achieving the same result. And `HasSize` seems to me like it could have a lot of overlap with `HasShape`.
> 
> This library isn't the venue for changing the Array API...

I know, and I wasn't suggesting anything like that.

Reply:
Is that not the practical effect of omission?